### DISCUSSION TITLE
Collection literals produce ambiguous call to methods accepting `char[]` and `string?`

### DISCUSSION THREAD

**Discussion by eduherminio** [dotnet/runtime#95548] [Q&A] [answered]
I'm exposing the following methods in my NuGet library.

```csharp
public ParsedFile(string path, char[] existingSeparator, bool ignoreEmptyItems = true) {}

public ParsedFile(string path, string? existingSeparator = null, bool ignoreEmptyItems = true) {}
```

But I just realized I can't call it using the new collection literals:

```csharp
new ParsedFile(InputFilePath, [ ';', ':' ]);
```

 or I get [CS0121](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0121) compiler error: ambiguous call.

![image](https://github.com/dotnet/runtime/assets/11148519/300b3492-264c-4aaf-bec8-d91c010552b3)

So I need to stick to invoking it like this:
```csharp
new ParsedFile(InputFilePath, new[] { ';', ':' })
```

----

I still want to allow users to define either a string as a separator (i.e. "//") or a number of different chars (i.e. `['.', ',', ';']`).

What's my best option, maintainability wise, when moving forward with the API?
Any way of making it compatible with collection literals?

Thanks 😀

**Comment by huoyaoyuan** [comment, marked as answer]
This is expected. See https://github.com/dotnet/roslyn/issues/70831 .
The language team is going to explore this in next version.
